python/ray/serve/examples/doc/snippet_model_ensemble.py

```diff
@@ -0,0 +1,57 @@
+import ray
+import time
+import asyncio
+
+from ray import serve
+
+
+@serve.deployment(num_replicas=5)
+def model_one(input_data):
+    print("Model 1 predict")
+    time.sleep(4)
+    return 1
+
+@serve.deployment(num_replicas=5)
+def model_two(input_data):
+    print("Model 2 predict")
+    time.sleep(4)
+    return 2
+
+@serve.deployment(max_concurrent_queries=10, route_prefix="/composed")
+class EnsembleModel:
+    def __init__(self):
+        self.model_one = model_one.get_handle()
+        self.model_two = model_two.get_handle()
+
+    async def __call__(self, input_data):
+        print("Call models concurrently, wait for both to finish")
+        tasks = [self.model_one.remote(data=input_data), self.model_two.remote(data=input_data)]
+        await asyncio.wait(tasks)
```

Comment:
each `.remote()` call returns a coroutine and `ray.get` can deal with them, similar to the case of https://discuss.ray.io/t/ray-serve-ray-serve-vs-regular-web-server-performance/4513/2

Can we try to skip this line and just do `ray.get` instead without introducing asyncio syntax in docs if not needed ? I think it will work too.

Reply:
@jiaodong it seems that unfortunately you missed the ensemble logic in this example. The examle you have provided here:
https://discuss.ray.io/t/ray-serve-ray-serve-vs-regular-web-server-performance/4513/2

does not perform ensemble prediction.

An ensemble model wants to run the predictions of the sub models in parallel and waits until both are done via async.
I need to emphasize "both" here, because you need asyncio.wait()  here to tell the event loop to wait until both are finished. This way the async ensemble model can give control back and handle the next input calls and again sends the predicitons to all sub models.

If you call ray.get(tasks)   this is immediately blocking and no async functionality of the ensemble model will be applied.
This means by using your suggestion, the model would wait 4 seconds with every call until the next call will be coming.
In the end 10 seconds of duration.

My proposed example runs asynchron  thx to await,on both tasks  and can serve  everything in 4 seconds this way.

-  if you use an async handle, I dont see a way how ray.get() would help there, since ray.get would need Obj refs.

If you find a way to perform the exact same ensemble use case as I have provided without asynchio.wait()  using only ray API  please tell me. I think this is not possible at the moment.
The only way would be scaling the ensemble model by spawning 5 replicas and using blocking ray.get().

But this example is all about async scaling of the ensemble model. In practice one could also scale it parrallel and use the asynch solution.

If I am wrong please tell me ;)

Reply:
-----------------------------------------
Maybe we should explain to the users, that there are two ways of scaling a Collecter or Ensemble or Composition deployment:

1. Async    via  await asyncio.wait(all_predictions_of_every_sub_model)
2. horizontal parallel process via num_replicas

These kind of Deplyoments are only Network I/O bound: just sending signals to all sub models and collecting/merging it.

These tasks can be scaled in both ways, maybe it is even better to combine the two to increase throughput.

Reply:
Thanks for the thorough explanation @frosk1 ! Let me try your example with a few possibilities and see how we can make it clear to user about the tradeoff and implications on doc.

Reply:
Thank you for the suggestion @jiaodong . I agree a 100%, forgot about the asyncio.gather methodology.